packages/compiler/src/server/serverlib.ts

```diff
@@ -347,6 +367,91 @@ export function createServer(host: ServerHost): Server {
     }
   }
 
+  async function renameFiles(params: RenameFilesParams): Promise<void> {
+    const firstFilePath = params.files[0];
+    if (!firstFilePath) {
+      return;
+    }
+
+    // Update cache for renamed folders.
+    for (const file of params.files) {
+      const oldFilePath = await fileService.getPath({ uri: file.oldUri });
+      const newFilePath = await fileService.getPath({ uri: file.newUri });
+      const isDirRename = !file.oldUri.endsWith(".tsp");
+
+      if (isDirRename) {
+        const files = await listAllFilesInDir(compilerHost, newFilePath);
+        for (const file of files) {
+          const oldFile = resolvePath(oldFilePath, file);
+
+          fileSystemCache.notify([
+            { uri: fileService.getURL(oldFile), type: FileChangeType.Deleted },
+          ]);
+        }
+      }
+    }
+
+    const mainFile = await compileService.getMainFileForDocument(
+      await fileService.getPath({ uri: firstFilePath.newUri }),
+    );
+
+    // Add this method to resolve timing issues between renamed files and `fs.stat`
+    // to prevent `fs.stat` from getting the files before modification.
+    // Currently the test requires a delay of 300ms
+    await new Promise((resolve) => setTimeout(resolve, 300));
+    const result = await compileService.compile({ uri: fileService.getURL(mainFile) });
+    if (!result) {
+      log({
+        level: "debug",
+        message: `The main tsp file '${mainFile}' compilation failed, please check the detailed compilation message`,
+      });
+      return;
+    }
+
+    for (const file of params.files) {
+      const oldFilePath = await fileService.getPath({ uri: file.oldUri });
+      const newFilePath = await fileService.getPath({ uri: file.newUri });
+      const isDirRename = !file.oldUri.endsWith(".tsp");
+
+      if (isDirRename) {
+        // Update cache for renamed files.
+        const files = await listAllFilesInDir(compilerHost, newFilePath);
+        for (const file of files) {
+          const oldFile = resolvePath(oldFilePath, file);
+
+          fileSystemCache.notify([
```

Comment:
we have already clean up the cache above, why do we need to clean the cache again here?